### PR TITLE
OP-213 add search field for opd records

### PR DESCRIPTION
**Pull request by mwithi**
See OP-213.

Add also USER column and related filter (like in Wolisso's project).

Paired with https://github.com/informatici/openhospital-core/pull/863.

### CODE REVIEW

File: bundle/language_en.properties [hunk 1]
```diff
@@ -1221,8 +1227,9 @@ angal.opd.reattendance.btn
 angal.opd.reattendance.txt                                                                             = Re-Attendance
 angal.opd.referral.txt                                                                                 = Referral
 angal.opd.referralto.txt                                                                               = Referred To
+angal.opd.searchbycodespressenter.border                                                               = Search by codes (press ENTER)
```
Comment (dbmalkovsky) [inline bundle/language_en.properties:1230]:
```suggestion
angal.opd.searchbycodespressenter.border                                                               = Search by Code (press ENTER)
```

File: bundle/language_en.properties [hunk 2]
```diff
@@ -1207,9 +1211,11 @@ angal.opd.opdnumberalreadyexist.msg
 angal.opd.opdnumbermustbeanumber.msg                                                                   = OPD No. must be a number.
 angal.opd.opdoutpatientdepartment.title                                                                = OPD Out Patient Department
 angal.opd.operation                                                                                    = Operation
+angal.opd.otherfilters.border                                                                          = Other filters
```
Comment (dbmalkovsky) [inline bundle/language_en.properties:1214]:
```suggestion
angal.opd.otherfilters.border                                                                          = Other Filters
```

**Comment by dbmalkovsky** [comment]
What about stacking the 3 entry fields  in a single column under the "Search by Code"?  That would use up some of the dead space at the bottom of the column and make it less wide

image

File: bundle/language_en.properties [hunk 3]
```diff
@@ -1192,6 +1195,7 @@ angal.opd.fullname.col
 angal.opd.insertavalidage.msg                                                                          = Insert a valid age.
 angal.opd.lastopdnote.txt                                                                              = Note
 angal.opd.lastopdvisitm.txt                                                                            = LAST OPD VISIT
+angal.opd.mypatient.btn                                                                                = My patients
```
Comment (dbmalkovsky) [inline bundle/language_en.properties:1198]:
```suggestion
angal.opd.mypatient.btn                                                                                = My Patients
```

**Comment by dbmalkovsky** [comment]
And maybe stack the Patient Type into a single column too.

File: bundle/language_en.properties (hunk 1 above)
Comment (mwithi) [reply bundle/language_en.properties:1230]: The plural was intentional (we have three different codes)
Is it bad English?

**Comment by mwithi** [comment]
> What about stacking the 3 entry fields in a single column under the "Search by Code"? That would use up some of the dead space at the bottom of the column and make it less wide
> 
> image

The width is given by the DiseaseBox, not the three fields, althought I don't know why in your screenshot it looks the opposite... Anyway I will try to see how it comes on my side.

**Comment by mwithi** [comment]
this is the result: as said, the problem is (at least for me) the diseasebox that need to be limited (with setMaximumSize) to crop very long diseases' descriptions. We need also to setPreferredSize, because from the biggest preffered size depends the whole container hierarchy, and so also the three new fields size.

image

File: bundle/language_en.properties (hunk 1 above)
Comment (dbmalkovsky) [reply bundle/language_en.properties:1230]: The search only uses one of the fields so one is searching on single code.

File: src/main/java/org/isf/opd/gui/OpdBrowser.java [hunk 4]
```diff
@@ -421,43 +437,125 @@ private JButton getJDeleteButton() {
...
+		searchCodesPanel.add(new JLabel(MessageBundle.getMessage("angal.common.patientID")));
+		searchCodesPanel.add(opdCodeFilter);
+		searchCodesPanel.add(progYearFilter);
+		searchCodesPanel.add(patientCodeFilter);
+		SpringUtilities.makeCompactGrid(searchCodesPanel, 2, 3, 5, 5, 5, 5);
+		return searchCodesPanel;
+	}
+
+	private JButton getResetButton() {
+		if(resetButton == null) {
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/gui/OpdBrowser.java:526]:
```suggestion
		if (resetButton == null) {
```

File: src/main/java/org/isf/opd/gui/OpdBrowser.java [hunk 5]
```diff
@@ -1038,4 +1118,132 @@ private JButton getFilterButton() {
 		return filterButton;
 	}
 
+	private char getGender() {
+		char sex;
+		if (radioAllGender.isSelected()) {
+			sex='A';
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/gui/OpdBrowser.java:1124]:
```suggestion
			sex = 'A';
```

File: src/main/java/org/isf/opd/gui/OpdBrowser.java [hunk 6]
```diff
@@ -1038,4 +1118,132 @@ private JButton getFilterButton() {
...
+		} else {
+			if (radioMale.isSelected()) {
+				sex='M';
+			} else {
+				sex='F';
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/gui/OpdBrowser.java:1129]:
```suggestion
				sex = 'F';
```

File: src/main/java/org/isf/opd/gui/OpdBrowser.java [hunk 7]
```diff
@@ -1038,4 +1118,132 @@ private JButton getFilterButton() {
...
+			}
+		}
+		return sex;
+	}
+	
+	private String getUser() {
+		if (!isSingleUser && radioMyPatients.isSelected())
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/gui/OpdBrowser.java:1136]:
```suggestion
		if (!isSingleUser && radioMyPatients.isSelected()) {
```

File: src/main/java/org/isf/opd/gui/OpdBrowser.java [hunk 8]
```diff
@@ -1038,4 +1118,132 @@ private JButton getFilterButton() {
...
+			return MainMenu.getUser().getUserName();
+		else
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/gui/OpdBrowser.java:1138]:
```suggestion
		} else {
```

File: src/main/java/org/isf/opd/gui/OpdBrowser.java [hunk 9]
```diff
@@ -1038,4 +1118,132 @@ private JButton getFilterButton() {
...
+			return null;
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/gui/OpdBrowser.java:1139]:
```suggestion
		return null;
```

File: src/main/java/org/isf/opd/gui/OpdBrowser.java [hunk 10]
```diff
@@ -1038,4 +1118,132 @@ private JButton getFilterButton() {
...
+	}
+
+	private char getPatientAttendance() {
+		if(radioAllPatiens.isSelected())
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/gui/OpdBrowser.java:1143]:
```suggestion
		if (radioAllPatiens.isSelected()) {
```

File: src/main/java/org/isf/opd/gui/OpdBrowser.java [hunk 11]
```diff
@@ -1038,4 +1118,132 @@ private JButton getFilterButton() {
...
+			return 'A';
+		else if(radioNewAttendance.isSelected())
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/gui/OpdBrowser.java:1145]:
```suggestion
		} else if (radioNewAttendance.isSelected()) {
```

File: src/main/java/org/isf/opd/gui/OpdBrowser.java [hunk 12]
```diff
@@ -1038,4 +1118,132 @@ private JButton getFilterButton() {
...
+			return 'N';
+		else
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/gui/OpdBrowser.java:1147]:
```suggestion
	}
```

File: src/main/java/org/isf/opd/gui/OpdBrowser.java [hunk 13]
```diff
@@ -1038,4 +1118,132 @@ private JButton getFilterButton() {
...
+			return 'R';
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/gui/OpdBrowser.java:1148]:
```suggestion
		return 'R';
```

File: src/main/java/org/isf/opd/gui/OpdBrowser.java [hunk 14]
```diff
@@ -605,50 +710,38 @@ public JPanel getSexPanel() {
...
-		}
-		return jSelectionWardPanel;
+	public JPanel getPatientTypePanel() {
+		JPanel patientTypePanel = new JPanel();
+		JLabel patientTypeLabel = new JLabel(MessageBundle.getMessage("angal.opd.patienttype.label"));
+		
+		ButtonGroup groupNewPatient = new ButtonGroup();
+		radioNewAttendance = new JRadioButton(MessageBundle.getMessage("angal.opd.new.btn"));
+		radioReAttendance = new JRadioButton(MessageBundle.getMessage("angal.opd.reattendance.btn"));
+		radioAllPatiens = new JRadioButton(MessageBundle.getMessage("angal.common.all.btn"));
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/gui/OpdBrowser.java:734]: FWIW, the properties that end in `.btn` are really for buttons in button panels like `Cancel`, `Save`, `Edit`, etc. and not for radio buttons.

File: src/main/java/org/isf/opd/gui/OpdBrowser.java [hunk 15]
```diff
@@ -605,50 +710,38 @@ public JPanel getSexPanel() {
...
-			radioa.setSelected(true);
-			group.add(radiom);
-			group.add(radiof);
-			group.add(radioa);
-			sexPanel.add(radioa);
-			sexPanel.add(radiom);
-			sexPanel.add(radiof);
+			radioMale = new JRadioButton(MessageBundle.getMessage("angal.common.male.btn"));
+			radioFemale = new JRadioButton(MessageBundle.getMessage("angal.common.female.btn"));
+			radioAllGender = new JRadioButton(MessageBundle.getMessage("angal.common.all.btn"));
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/gui/OpdBrowser.java:715]: See note at line 734.

File: src/main/java/org/isf/opd/gui/OpdBrowser.java [hunk 16]
```diff
@@ -469,21 +567,29 @@ private JLabel getRowCounter() {
...
 			SpringUtilities.makeCompactGrid(dateFilterPanel, 2, 2, 5, 5, 5, 5);
 		}
 		return dateFilterPanel;
 	}
 
+	private void resetDates() {
+		if (!GeneralData.ENHANCEDSEARCH) {
+			dateFrom.setDate(LocalDate.now().minusWeeks(1));
+		} else {
+			dateFrom.setDate(LocalDate.now());
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/gui/OpdBrowser.java:588]:
```suggestion
			dateFrom.setDate(TimeTools.getNow()));
```

We had decided to replace `LocalDate.now()` with '`TimeTools.getNow()` didn't we?

File: src/main/java/org/isf/opd/gui/OpdBrowser.java (hunk 14 above)
Comment (mwithi) [reply src/main/java/org/isf/opd/gui/OpdBrowser.java:734]: I found it this way in other parts, so I did the same, should we change that?

image

File: src/main/java/org/isf/opd/gui/OpdBrowser.java (hunk 16 above)
Comment (mwithi) [reply src/main/java/org/isf/opd/gui/OpdBrowser.java:588]: yep! Thanks!

File: src/main/java/org/isf/opd/gui/OpdBrowser.java (hunk 14 above)
Comment (dbmalkovsky) [reply src/main/java/org/isf/opd/gui/OpdBrowser.java:734]: At some point maybe but let's ignore it for now 😃

**Comment by mwithi** [comment]
Ok, so the proposed graphic layout is ok for now?

**Comment by dbmalkovsky** [comment]
yes the layout is ok for now

**Comment by mwithi** [comment]
> yes the layout is ok for now

Good, thanks a lot! I proceed with the doc

File: src/main/java/org/isf/opd/gui/OpdBrowser.java [hunk 17]
```diff
@@ -421,43 +437,125 @@ private JButton getJDeleteButton() {
...
+		searchCodesPanel.add(opdCodeFilter);
+		searchCodesPanel.add(progYearFilter);
+		searchCodesPanel.add(patientCodeFilter);
+		SpringUtilities.makeCompactGrid(searchCodesPanel, 2, 3, 5, 5, 5, 5);
+		return searchCodesPanel;
+	}
+
+	private JButton getResetButton() {
+		if (resetButton == null) {
+			resetButton = new JButton("Reset");
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/gui/OpdBrowser.java:527]: shouldn't this be a property string and not a hard code literal?

File: src/main/java/org/isf/opd/gui/OpdBrowser.java [hunk 18]
```diff
@@ -469,21 +567,29 @@ private JLabel getRowCounter() {
 	private Component getDateFilterPanel() {
 		if (dateFilterPanel == null) {
 			dateFilterPanel = new JPanel(new SpringLayout());
+			
+			dateFrom = new GoodDateChooser(LocalDate.now());
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/gui/OpdBrowser.java:571]:
```suggestion
			dateFrom = new GoodDateChooser(TimeTools.getnow());
```
```

File: src/main/java/org/isf/opd/gui/OpdBrowser.java [hunk 19]
```diff
@@ -469,21 +567,29 @@ private JLabel getRowCounter() {
...
+			dateTo = new GoodDateChooser(LocalDate.now());
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/gui/OpdBrowser.java:572]:
```suggestion
			dateTo = new GoodDateChooser(TimeTools.getnow());
```

File: src/main/java/org/isf/opd/gui/OpdBrowser.java [hunk 20]
```diff
@@ -469,21 +567,29 @@ private JLabel getRowCounter() {
...
-			dateTo = new GoodDateChooser(LocalDate.now());
 			dateFilterPanel.add(dateTo);
 			SpringUtilities.makeCompactGrid(dateFilterPanel, 2, 2, 5, 5, 5, 5);
 		}
 		return dateFilterPanel;
 	}
 
+	private void resetDates() {
+		if (!GeneralData.ENHANCEDSEARCH) {
+			dateFrom.setDate(LocalDate.now().minusWeeks(1));
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/gui/OpdBrowser.java:586]:
```suggestion
			dateFrom.setDate(TimeTools.getnow()minusWeeks(1));
```

File: src/main/java/org/isf/opd/gui/OpdBrowser.java [hunk 21]
```diff
@@ -1038,4 +1118,132 @@ private JButton getFilterButton() {
...
+		}
+		return null;
+	}
+
+	private char getPatientAttendance() {
+		if (radioAllPatiens.isSelected()) {
+			return 'A';
+		} else if (radioNewAttendance.isSelected()) {
+			return 'N';
+	}
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/gui/OpdBrowser.java:1147]: formatting of the brace looks like it is indented too much to the left

File: src/main/java/org/isf/opd/gui/OpdBrowser.java (hunk 17 above)
Comment (mwithi) [reply src/main/java/org/isf/opd/gui/OpdBrowser.java:527]: Good catch! Let me fix that! Thanks!

File: src/main/java/org/isf/opd/gui/OpdBrowser.java [hunk 22]
```diff
@@ -469,21 +567,29 @@ private JLabel getRowCounter() {
...
 		return dateFilterPanel;
 	}
 
+	private void resetDates() {
+		if (!GeneralData.ENHANCEDSEARCH) {
+			dateFrom.setDate(LocalDate.now().minusWeeks(1));
+		} else {
+			dateFrom.setDate(TimeTools.getNow()));
+		}
+		dateTo.setDate(LocalDate.now());
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/opd/gui/OpdBrowser.java:590]:
```suggestion
		dateTo.setDate(TimeTools.getNow());
```

**Review by dbmalkovsky — APPROVED** [review]
LGTM when the latest code review comments are done.

File: src/main/java/org/isf/opd/gui/OpdBrowser.java (hunk 20 above)
Comment (mwithi) [reply src/main/java/org/isf/opd/gui/OpdBrowser.java]: getNow() is returning a LocalDateTime, not a LocalDate, so it is not applicable in this case.
same for the GoodDateChooser constructor parameter

image

File: src/main/java/org/isf/opd/gui/OpdBrowser.java (hunk 21 above)
Comment (mwithi) [reply src/main/java/org/isf/opd/gui/OpdBrowser.java:1147]: done, thanks!

File: src/main/java/org/isf/opd/gui/OpdBrowser.java (hunk 20 above)
Comment (dbmalkovsky) [reply src/main/java/org/isf/opd/gui/OpdBrowser.java:586]: you are right; I'm just moving too fast; ignore any others like that.

File: src/main/java/org/isf/opd/gui/OpdBrowser.java (hunk 22 above)
Comment (dbmalkovsky) [reply src/main/java/org/isf/opd/gui/OpdBrowser.java:590]: Ignore this change too. ☹️

File: src/main/java/org/isf/opd/gui/OpdBrowser.java (hunk 20 above)
Comment (mwithi) [reply src/main/java/org/isf/opd/gui/OpdBrowser.java:586]: ok, no prob! Thanks!

PS: we should think about something like getDateNow() and getDateTimeNow() ?

File: src/main/java/org/isf/opd/gui/OpdBrowser.java (hunk 22 above)
Comment (mwithi) [reply src/main/java/org/isf/opd/gui/OpdBrowser.java:590]: ok, no prob! Thanks!

**Pull request merged**